ads/google/utils.js

```diff
@@ -0,0 +1,147 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {user} from '../../src/log';
+
+/**
+ * Given the amp-ad data attribute containing the multi-size dimensions, and a
+ * set of primary dimensions, this function will return all valid multi-size
+ * [width, height] pairs in an array.
+ *
+ * @param {string} multiSizeDataStr The amp-ad data attribute containing the
+ *   multi-size dimensions.
+ * @param {number} primaryWidth The primary width of the ad slot.
+ * @param {number} primaryHeight The primary height of the ad slot.
+ * @param {!Array<Array<number>>=} opt_dimensions An array into which to put
+ *   the multi-size dimensions.
+ * @param {boolean} multiSizeValidation A flag that if set to anything other
+ *   than false will enforce the rule that ensures multi-size dimensions are no
+ *   less than 2/3rds of their primary dimension counterpart.
+ * @return {!Array<Array<number>>} An array of arrays, the latter of which
+ *   contains the width and height as its sole elements (in that order).
+ */
+export function getMultiSizeDimensions(
+    multiSizeDataStr,
+    primaryWidth,
+    primaryHeight,
+    multiSizeValidation,
+    opt_dimensions) {
+
+  const dimensions = opt_dimensions || [];
+  const arrayOfSizeStrs = multiSizeDataStr.split(',');
+
+  arrayOfSizeStrs.forEach(sizeStr => {
+
+    const size = sizeStr.split('x');
+
+    // Make sure that each size is specified in the form val1xval2.
+    if (size.length != 2) {
+      user().error('AMP-AD', `Invalid multi-size data format '${sizeStr}'.`);
+      return;
+    }
+
+    const width = Number(size[0]);
+    const height = Number(size[1]);
+
+    // Make sure that both dimensions given are numbers.
+    if (!validateDimensions(width, height,
+          w => isNaN(Number(w)),
+          h => isNaN(Number(h)),
+          ({badDim, badVal}) =>
+          `Invalid ${badDim} of ${badVal} given for secondary size.`)) {
+      return;
+    }
+
+    // Check that secondary size is not larger than primary size.
+    if (!validateDimensions(width, height,
+          w => w > primaryWidth,
+          h => h > primaryHeight,
+          ({badDim, badVal}) => `Secondary ${badDim} ${badVal} ` +
+          `can't be larger than the primary ${badDim}.`)) {
+      return;
+    }
+
+    // Check that if multi-size-validation is on, that the secondary sizes
+    // are at least minRatio of the primary size.
+    const validate = multiSizeValidation || true;
```

Comment:
Ah, this must be why I used `'true'` rather than `true` in my initial implementation. The multi-size-validation data attribute is true by default, so I just need to check whether the slot supplied a value of `'false'` or an explicit `'true'`.